cmd/sriov-network-config-daemon/service.go

```diff
@@ -0,0 +1,166 @@
+			SyncStatus:    "Failed",
+			LastSyncError: fmt.Sprintf("failed to read list of supported nic ids: %v", err),
+		}
+		err = systemd.WriteSriovResult(sriovResult)
+		if err != nil {
+			glog.Errorf("sriov-config-service failed to write sriov result file with content %v error: %v", *sriovResult, err)
+			return
+		}
+		return
+	}
+	sriovv1.InitNicIDMapFromList(supportedNicIds)
+
+	nodeStateSpec, err := systemd.ReadConfFile()
+	if err != nil {
+		if _, err := os.Stat(utils.SriovSwitchDevConfPath); !errors.Is(err, os.ErrNotExist) {
+			glog.Errorf("failed to read the sriov configuration file in path %s: %v", utils.SriovSwitchDevConfPath, err)
+			sriovResult := &systemd.SriovResult{
+				SyncStatus:    "Failed",
+				LastSyncError: fmt.Sprintf("failed to read the sriov configuration file in path %s: %v", utils.SriovSwitchDevConfPath, err),
+			}
+			err = systemd.WriteSriovResult(sriovResult)
+			if err != nil {
+				glog.Errorf("sriov-config-service failed to write sriov result file with content %v error: %v", *sriovResult, err)
+				return
+			}
+		}
+
+		nodeStateSpec = &systemd.SriovConfig{
+			Spec:            sriovv1.SriovNetworkNodeStateSpec{},
+			UnsupportedNics: false,
+			PlatformType:    utils.Baremetal,
+		}
+	}
+
+	glog.V(2).Infof("sriov-config-service read config: %v", nodeStateSpec)
+
+	// Load kernel modules
+	hostManager := host.NewHostManager(true)
+	_, err = hostManager.TryEnableRdma()
+	if err != nil {
+		glog.Warningf("failed to enable RDMA: %v", err)
+	}
+	hostManager.TryEnableTun()
+	hostManager.TryEnableVhostNet()
+
+	var configPlugin plugin.VendorPlugin
+	var ifaceStatuses []sriovv1.InterfaceExt
+	if nodeStateSpec.PlatformType == utils.Baremetal {
+		// Bare metal support
+		ifaceStatuses, err = utils.DiscoverSriovDevices(nodeStateSpec.UnsupportedNics)
+		if err != nil {
+			glog.V(0).Infof("sriov-config-service failed %v", err)
+			return
+		}
+
+		// Create the generic plugin
+		configPlugin, err = generic.NewGenericPlugin(true)
+		if err != nil {
+			glog.Errorf("sriov-config-service failed to create generic plugin %v", err)
+			return
+		}
+	} else if nodeStateSpec.PlatformType == utils.VirtualOpenStack {
+		// Openstack support
+		metaData, networkData, err := utils.GetOpenstackData(false)
+		if err != nil {
+			glog.Errorf("RunOnce(): failed to read OpenStack data: %v", err)
+			return
+		}
+
+		openStackDevicesInfo, err := utils.CreateOpenstackDevicesInfo(metaData, networkData)
+		if err != nil {
+			glog.Errorf("RunOnce(): failed to read OpenStack data: %v", err)
+			return
+		}
+
+		ifaceStatuses, err = utils.DiscoverSriovDevicesVirtual(openStackDevicesInfo)
+		if err != nil {
+			glog.Errorf("RunOnce(): failed to read OpenStack data: %v", err)
+			return
+		}
+	}
+
+	nodeState := &sriovv1.SriovNetworkNodeState{
+		Spec:   nodeStateSpec.Spec,
+		Status: sriovv1.SriovNetworkNodeStateStatus{Interfaces: ifaceStatuses},
+	}
+
+	_, _, err = configPlugin.OnNodeStateChange(nodeState)
```

Comment:
This is going to panic in case of `nodeStateSpec.PlatformType == utils.VirtualOpenStack`. 
Is it missing something like
```
configPlugin = virtualplugin.NewVirtualPlugin(true)
```
?

Reply:
right sorry I didn't test this one yet on OpenStack :(